pkg/types/validation/installconfig.go

```diff
@@ -99,6 +102,122 @@ func ValidateInstallConfig(c *types.InstallConfig, openStackValidValuesFetcher o
 	if _, ok := validPublishingStrategies[c.Publish]; !ok {
 		allErrs = append(allErrs, field.NotSupported(field.NewPath("publish"), c.Publish, validPublishingStrategyValues))
 	}
+
+	return allErrs
+}
+
+// IPAddressTypeByField is a map of field path to whether they request IPv4 or IPv6.
+type IPAddressTypeByField map[string]struct{ IPv4, IPv6 bool }
+
+// IPByField is a map of field path to the net.IPs in sorted order.
+type IPByField map[string][]net.IP
+
+// InferIPVersionFromInstallConfig infers the user's desired ip version from the networking config.
+// Presence field names match the field path of the struct within the Networking type. This function
+// assumes a valid install config.
+func InferIPVersionFromInstallConfig(n *types.Networking) (hasIPv4, hasIPv6 bool, presence IPAddressTypeByField, addresses IPByField) {
+	if n == nil {
+		return
+	}
+	addresses = make(IPByField)
+	for _, network := range n.MachineNetwork {
+		addresses["machineNetwork"] = append(addresses["machineNetwork"], network.CIDR.IP)
+	}
+	for _, network := range n.ServiceNetwork {
+		addresses["serviceNetwork"] = append(addresses["serviceNetwork"], network.IP)
+	}
+	for _, network := range n.ClusterNetwork {
+		addresses["clusterNetwork"] = append(addresses["clusterNetwork"], network.CIDR.IP)
+	}
+	presence = make(IPAddressTypeByField)
+	for k, ips := range addresses {
+		sort.Slice(ips, func(i, j int) bool { return bytes.Compare(ips[i], ips[j]) < 0 })
+		for _, ip := range ips {
+			has := presence[k]
+			if ip.To4() != nil {
+				has.IPv4 = true
+				hasIPv4 = true
+			} else {
+				has.IPv6 = true
+				hasIPv6 = true
+			}
+			presence[k] = has
+		}
+	}
+	return
+}
+
+func ipSliceToStrings(ips []net.IP) []string {
+	var s []string
+	for _, ip := range ips {
+		s = append(s, ip.String())
+	}
+	return s
+}
+
+func ipnetworksToStrings(networks []ipnet.IPNet) []string {
+	var diag []string
+	for _, sn := range networks {
+		diag = append(diag, sn.String())
+	}
+	return diag
+}
+
+// validateNetworkingIPVersion checks parameters for consistency when the user
+// requests single-stack IPv6 or dual-stack modes.
+func validateNetworkingIPVersion(c *types.InstallConfig) field.ErrorList {
```

Comment:
is it possible to declare more closer inputs, ie Networking and platform name ?

Reply:
Yes, will do.